pgroll-internals.mdx

```diff
@@ -0,0 +1,576 @@
+published: true
+slug: pgroll-internals
+ogImage: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/introducing-pgroll.jpg
+---
+
+## pgroll Internals
+
+At the start of October we released [pgroll](https://github.com/xataio/pgroll), an open source tool for zero-downtime, reversible schema migrations for Postgres. We blogged about it at the time here: [Introducing pgroll: zero-downtime, reversible, schema migrations for Postgres](https://xata.io/blog/pgroll-schema-migrations-postgres).
+
+We've been hard at work on `pgroll` in the couple of months since and we thought now would be a good time to take a deeper dive into `pgroll` and how it does what it does.
+
+## A brief recap
+
+`pgroll` is an open source schema migration tool that takes a different approach, using the [expand/contract pattern](https://openpracticelibrary.com/practice/expand-and-contract-pattern/), to solve some of the problems associated with schema migrations:
+
+- **Multiple schema versions**: `pgroll` keeps two versions of your schema active at the same time during a migration; the old schema and the new. This means that old versions of client applications can co-exist with new versions of client applications during an application rollout because each version of the application sees the version of the database schema that it expects to see.
+- **Lock free migrations**: A careful approach to locking to ensure that tables don't get locked for long periods during a migration.
+- **Easy rollbacks**: Rollbacks become trivial because `pgroll` maintains two versions of your database schema during a migration. A rollback means simply deleting the new version of the schema. The old version never went away so old versions of client applications never see any disruption.
+- **Data backfilling**: When old and new versions of a schema co-exist during a migration `pgroll` ensures that any data written to the old version of the database schema is reflected in the new version and vice-versa. This ensures that old and new versions of client applications can run side-by-side indefinitely until the migration is declared complete.
+
+## A step by step example
+
+Let's see how `pgroll` works by walking through an example migration and seeing what `pgroll` does at each step of the migration process.
+
+### Initialization
+
+`pgroll` needs to store its own internal state somewhere in the target Postgres database. Initializing `pgroll` configures this store and makes `pgroll` ready for first use:
+
+```
+pgroll init
+```
+
+You should see a success message indicating that `pgroll` has been configured.
+
+> What data does `pgroll` store?
+> `pgroll` stores its data in the `pgroll` schema. In this schema it creates:
+>
+> - A `migrations` table containing the version history for each schema in the database
+> - Functions to capture the current database schema for a given schema name
+> - Triggers to capture DDL statements run outside of `pgroll` migrations
```

Comment:
Definitely a fly-by and unimportant comment and could very well be just me, but for some reason this format looks off to me having the Q&A side by side vs. top to bottom. I think it's mostly because of the limited space.

Reply:
This is the one quote box I can't get to look right. I've tried various things but the renderer always puts the list to one side.